Add support for disabling beatmap preview buttons

frenzibyte opened the pull request:
This PR is adding support for disabling play buttons programmatically, which will be used in an upcoming PR.

I'm not entirely sure whether hiding the button completely when disabled is the way, but that's how it's done osu!web-side.

Also in case anyone wonders, the change in `PreviewButton` was because `PlayButton` doesn't do anything in `OnClick` now, `PlayButton.Button` does, and either way that `Click()` method really shouldn't exist, it's way hacky to the point clicks aren't propagated to the drawable's children, if for testing, `InputManager` is more preferable to use.

frenzibyte committed:
Add support for disabling beatmap listing play buttons

frenzibyte committed:
Fix preview button sending wrong click events

frenzibyte committed:
Add simple test coverage

frenzibyte committed:
Update `PreviewButton` with play button's enabled state

frenzibyte committed:
Disable play button on null beatmap set as a use case

frenzibyte commented on osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:
```diff
@@ -74,6 +74,13 @@ private void load(OsuColour colours, OverlayColourProvider colourProvider)
         }
 
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            playButton.Enabled.BindValueChanged(e => Enabled.Value = e.NewValue, true);
```
well, that's the design of `PreviewButton`, it creates a `PlayButton` inside of it, `PlayButton` is the one managing over the beatmap set's preview track, `PreviewButton` only adds to the look of the button. don't think it's feasible to change the direction of propagation without rewriting `PreviewButton` all over.

bdach replied:
no, that's not quite the same. the preview track and the playing bindable are re-exposed in `PreviewButton`. it's composition. if it were possible to proxy `Enabled`, i'd recommend that, but that doesn't seem feasible as the preview button already has its own `Enabled`. so it'd have to be declared with `new` which i don't even know if it'll behave correctly since shadowing is confusing.

speaking of which, given this direction of propagation, how do you even disable a `PreviewButton`? setting `Enabled` on the preview doesn't propagate to the play button, and you can't disable the inner button from outside because it's private.

frenzibyte replied:
All good points, does this commit satisfy what you wanted https://github.com/ppy/osu/commit/1734a889b838023ca310bfc3fd5378dc1f22fce2, otherwise not sure I can come up with anything else simple to do, helping would be appreciated.

peppy replied:
The direction you have on that commit seems.. better? see comments i left on it.

frenzibyte commented on osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:
```diff
@@ -64,6 +69,17 @@ public PlayButton(BeatmapSetInfo setInfo = null)
+        private void updateEnabledState()
+        {
+            button.Enabled.Value = BeatmapSet != null;
+        }
```
well, this isn't actually meant to be modifiable programmatically, only done so in test scene by straight force, so it doesn't matter by much for now.

main PR uses the internal button enabled state this way, if you may ask:
https://github.com/ppy/osu/blob/3f16acaf82cfcd8ad911997bdc1d42eaf046f361/osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs#L84-L99

bdach replied:
that is all entirely backwards. all of it. `Enabled` is *public*. it needs to work the same way 99% of all other components work. so it *must* be settable programmatically and it *must* work reliably.

bdach commented on osu.Game.Tests/Visual/Online/TestSceneDirectPanel.cs:
```diff
@@ -125,13 +130,38 @@ private void load(RulesetStore rulesets)
+            {
+                base.LoadComplete();
+
+                PlayButton.ChildrenOfType<ClickableContainer>().Single().Enabled.Value = false;
```
this also needs to go away. it should be no more difficult than directly setting `Enabled` on the button to test this.

frenzibyte committed:
Reverse disabled state propagation

bdach commented on osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:
```diff
@@ -39,88 +46,50 @@ public BeatmapSetInfo BeatmapSet
+            playing.Disabled = disabled;
         }
 
+        public void TogglePlaying() => playing.Toggle();
```
what's the point? `Playing` (the public one) is... well, public.

frenzibyte replied:
that would make `Playing.Disabled` mutable, which is not wanted

bdach commented on osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:
```diff
@@ -74,6 +77,20 @@ private void load(OsuColour colours, OverlayColourProvider colourProvider)
+        {
+            base.LoadComplete();
+
+            Playing.BindDisabledChanged(_ => updateEnabledState());
+            Enabled.BindValueChanged(_ => updateEnabledState());
```
so given that `Enabled` sets `playing.Disabled`, are these both going to fire? this is still way too complicated for how it should be. it shouldn't be this hard to disable a button. i have no clue why that boolean value has to be duplicated in 2 places and moved all over.

frenzibyte replied:
I've just realized `PreviewButton` being a clickable container and exposing an `Enabled` bindable is just useless, so I removed the inheritance and it seems to have improved readability pretty well.

frenzibyte committed:
Remove clickable container inheritance for simplicity

frenzibyte committed:
Remove unintended hover click sounds from `PlayButton`'s button

bdach commented on osu.Game/Overlays/BeatmapSet/Buttons/PreviewButton.cs:
```diff
@@ -61,9 +61,9 @@ public PreviewButton()
                     Origin = Anchor.Centre,
                     Size = new Vector2(18),
                 },
+                new HoverClickSounds()
```
why go to the trouble of not inheriting `OsuClickableContainer` anymore if you're just going to jam `HoverClickSounds` back in, which is like half of the point of that class?

i swear this is all because of the baffling insistence on storing the button's enabled state in `playing.Disabled` instead of a separate `Enabled` bindable, which is both confusing and just doesn't fit the general design of everything around it (which is apparent by re-implementing button disable logic all over the place). i can't see a reasonable justification for it at all.

frenzibyte replied:
no, it's literally because I can't find any way to do this cleanly without extra refactoring, `PlayButton` is disabling itself very innocently on some conditions, and with the current layout, `PreviewButton`, which is containing `PlayButton`, has to be disabled based on whether the play button is.

i've removed inheritance to `OsuClickableContainer` to remove exposing `Enabled` bindable and get rid of hiding properties and `base.Enabled` stuff to improve readability (and even get rid of public `Action` event while there, `OsuClickableContainer` isn't even correct for this type of button).

honestly at this point i'm just thinking about making `PreviewButton` inherit `PlayButton` and expose some method to create extra content around the button, it's the only thing that would make sense and would get this over with.

frenzibyte committed:
Let `PlayButton.Enabled` contain the overall state and expose `AlwaysDisabled`

frenzibyte committed:
Allow `PlayButton`'s content to be customized

I think `PlayButton` should be renamed as well, not sure what name to give though, staying at old name for now.

frenzibyte committed:
Refactor `PreviewButton` into inheriting `PlayButton` directly instead

frenzibyte committed:
Fix play button not showing as hovered when re-enabled back

frenzibyte committed:
Remove leftover value change callback

frenzibyte committed:
Invert to `Disabled` and make `AlwaysDisabled` bindable

Feels better using `Disabled` over `Enabled` for consistency with the `AlwaysDisabled` bindable.

Also made `AlwaysDisabled` a bindable for simplifying code (no need for backing field)

peppy commented on osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:
```diff
@@ -166,5 +151,89 @@ private void attemptStart()
+                base.OnHoverLost(e);
+            }
+
+            private void updateDisplay()
```
I'd probably make this a `virtual protected` to avoid duplicating all the calling logic of this function (it's being bound in three different events). Make sure to attach xmldoc mentioning it will be called on hover or disabled state changes.

In fact, if you send in hovered and disabled as argument you may be able to keep the disabled bindable private.

frenzibyte replied:
Well, one problem is that hovering over the icon button is different from hovering over `HeaderPlayButton`, so the `virtual protected` method would have to accept hovering state of both the `PlayButton` and the icon button, which sounds kinda too much from just seperating them.

peppy commented on osu.Game/Overlays/BeatmapListing/Panels/PlayButton.cs:
```diff
@@ -39,88 +49,63 @@ public BeatmapSetInfo BeatmapSet
         }
 
-        protected override bool OnHover(HoverEvent e)
+        private void updateEnabledState()
```
Move to end of file with other private methods.

frenzibyte committed:
Move `updateEnabledState()` method to end of file

peppy committed:
Merge branch 'master' into play-button-disable-support

peppy committed:
Update some naming

peppy commented:
I've changed some naming to read better, but right now this is a regression in code quality, to the point I can't easily follow what's going on. At very least, the bindable logic between `PlayButton` and its nested class needs to be removed. Just replace it with a method or two non-bindable properties.

frenzibyte commented:
I agree bindables can make things hard to follow in some cases, will take a look at that.

frenzibyte commented:
Just to be clear, are you referring to `ShouldDisplay` and `Playing`? as I'm thinking about:
 - a `protected bool` property for each of them.
 - a virtual `UpdateDisplay` method on the main `PlayButton` class called when any of the properties are changed.
 - an `UpdateState` method on the `PlayButton.Button` nested class called when any of the properties are changed.

peppy commented:
I'm pretty sure you still want the outwards facing ones as bindables? Aren't they being used in such a fashion? My complaint was about the communication between `PlayButton` and `PlayButton.Button`.

smoogipoo commented:
Going to close due to inactivity + going to require a herculean effort to merge due to the panels having since been rewritten. Best start off fresh.
